src/charm.py

```diff
@@ -100,23 +110,86 @@ def _on_config_changed(self, event):
         self.cluster_provider.publish_configs(mimir_config)
 
+    def create_minio_buckets(self, conn: dict, bucket_names: list):
+        """Create Minio buckets."""
+        try:
+            client = Minio(
+                endpoint=f'{conn["service"]}.{conn["namespace"]}.svc.cluster.local:{conn["port"]}',
+                access_key=conn["access_key"],
+                secret_key=conn["secret_key"],
+                secure=conn["secure"],
+            )
+            for bucket in bucket_names:
+                found = client.bucket_exists(bucket)
+                if not found:
+                    client.make_bucket(bucket)
+                else:
+                    logger.info("Bucket already exists")
+        except S3Error:
+            logger.error("Error creating S3 buckets")
+
+    def _parse_s3_data(self, s3_data):
+        data_passed_dict = dict(item.split(": ") for item in s3_data.split("\n") if item)
+        data_passed_dict["secure"] = data_passed_dict.get("secure", "").lower() == "true"
+
+        return {
+            "access_key": data_passed_dict.get("access-key", ""),
+            "namespace": data_passed_dict.get("namespace", ""),
+            "port": data_passed_dict.get("port", ""),
+            "secret_key": data_passed_dict.get("secret-key", ""),
+            "secure": data_passed_dict.get("secure", ""),
+            "service": data_passed_dict.get("service", ""),
+        }
+
     def _on_mimir_cluster_changed(self, _):
-        if self.coordinator.is_coherent():
-            logger.info("mimir deployment coherent: publishing configs")
-            self.publish_config()
-        else:
-            logger.warning("this mimir deployment is incoherent")
+        if not self.coordinator.is_coherent():
+            logger.warning("Incoherent deployment: Some required Mimir roles are missing.")
+            return
+        self._process_s3_relation()
+        if not self._s3_storage_data and self.coordinator.is_scaled():
+            logger.warning("Filesystem storage cannot be used with replicated mimir workers")
+            return
+        self.publish_config()
+
+    def _process_s3_relation(self):
+        s3_relation = self.model.get_relation("s3")
+        if s3_relation and s3_relation.app:
+            data_passed = s3_relation.data.get(s3_relation.app, {}).get("data")
+            if data_passed:
+                s3_parsed_data = self._parse_s3_data(data_passed)
+                self.create_minio_buckets(s3_parsed_data, ["mimir"])
+                self._s3_storage_data = s3_parsed_data
+                return
+        self._s3_storage_data = None
+
+    def _on_s3_created(self, _):
+        s3_relation = self.model.get_relation("s3")
+        if s3_relation:
+            s3_relation.data[self.model.app]["_supported_versions"] = "- v1"
+
+    def _on_s3_broken(self, _):
+        if not self.coordinator.is_coherent():
+            logger.warning("Incoherent deployment: Some required Mimir roles are missing.")
```

Comment:
From past experience it's best when status is set in a centralized way.
I wonder if [collect_status](https://github.com/canonical/operator/pull/954) would work here.

Ah I see you use it. The thing with collect_status is that it's an "all in" solution. These statuses here would have no effect now.

Reply:
Moved the check inside the collect_status